Renesas SmartBond(tm) power management

kasjer opened the pull request:
## Power management
This PR introduces power management to SmartBond(tm) family.
With all subsequent changes MCU can enter low power mode where CPU is powered down, GPIOs are latched unused power domains are powered down to reduce power consumption.

First thing needed to accomplished this is usage of TIMER2 which can run event when CPU is powered off. When this timer is used Cortex Tick timer is not needed and CPU can be powered off.

- Standby mode handling added (in standby mode CPU is powered off, while RAM is preserved, `sleep.S` provides code to that restores system state when CPU is waken up)
- This adds timer driver for Renesas SmartBond(tm) family.
It uses TIMER2 block which is in PD_TIM power domain so it can work even if ARM core is disabled, thus can work as a sleep timer.

- GPIOs are latched before CPU is powered down and PD_COM domain is also powered off during sleep.
- GPIOs functionality is restored when system wakes up. When #61844 is merged GPIOs changes will also be able to wake platform.
- DC/DC converter is can be configured to be used instead of LDO

### How to test
There is no separate application to test that system actually enters standby mode.
To test it's possibly to use standard **blinky** example with DA1469X-DK_PRO kit that has blue LED on the mother board. LED is lit when power consumption of MCU is above certain threshold detected by OpAMPs.
When blinky sample is run without power management code. Blue LED is lit all the time.
When power management (CONFIG_PM=y) is enabled with code in this PR blinky example will enter standby mode. When standby mode is entered and normal RED LED is off power consumption drops to the level that let BLUE LED to be turned off.

andrzej-kaczmarek committed:
drivers: timer: Add timer driver to Renesas SmartBond(tm)

This adds timer driver for Renesas SmartBond(tm) family.
It uses TIMER2 block which is in PD_TIM power domain so it can work even
if ARM core is disabled, thus can work as a sleep timer.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>
Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

andrzej-kaczmarek committed:
soc: arm: smartbond: Add support for extended sleep

This enabled extended sleep for Renesas SmartBond(tm).

Extended sleep is low power mode where ARM core is powered off and can
be woken up by PDC. This is default sleep mode when CONFIG_PM is
enabled.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>
Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

andrzej-kaczmarek committed:
soc: arm: smartbond: Always select PLATFORM_SPECIFIC_INIT

Platform specific init is needed once power management is introduced.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
soc: arm: smartbond: Enable cache retainability in sleep

This enables cache retainability while in sleep so there's no penalty
when executing from QSPI after wakeup.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
drivers: gpio: smartbond: Add GPIO latching for PM

This adds automatic GPIO latching before going to extended sleep and
restoring state after wakeup.

Mode and state for each pin is stored, then ports are latched to retain
state when PD_COM is disabled during sleep. On wakeup mode and state for
each pin is restored and ports are unlatched to make it work again.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>
Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

andrzej-kaczmarek commented on drivers/timer/smartbond_timer.c:
```diff
@@ -0,0 +1,157 @@
+
+#define COUNTER_SPAN BIT(24)
+#define CYC_PER_TICK (sys_clock_hw_cycles_per_sec() / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+#define MAX_TICKS ((COUNTER_SPAN / 2 - CYC_PER_TICK) / CYC_PER_TICK)
```
I think the max_ticks should be adjusted depending on watchdog settings, i.e. watchdog is always enabled after entering extended sleep so we need to make sure that we'll wake up before it expires

kasjer replied:
While I tried to add this functionality I'm not sure it's needed.
If CONFIG_WDT_SMARTBOND is not set. power.c will disable watchdog after wakeup.
If watchdog is configured. Whatever task is feeding watchdog must wakeup before watchdog timer expires.
If task/work or something else is not armed to be ready before watchdog bites, even if platform is woken from sleep responsible task will not be triggered and watchdog timeout will expire anyway.
Unless I'm missing something...

kasjer replied:
Now I see.
When watchdog is disabled it will start when PD_SYS is powered off and sleep time should not exceed this time.
Which can be RCX or RC32K based so it should be computed.

andrzej-kaczmarek replied:
correct, so code should be only enabled if CONFIG_PM is enabled to prevent watchdog from expiring during extended sleep period. as you said, in other case something should feed watchdog before it expires.

andrzej-kaczmarek commented:
@kasjer fixes look good, please squash them with appropriate commits

kasjer commented:
rebased without changes, maybe twister will work this time

gmarull commented on soc/arm/renesas_smartbond/da1469x/soc.c:
```diff
@@ -26,18 +30,147 @@
+	POWER_CTRL_REG_SET(VDD_LEVEL, vdd_level);
+}
+
+#if defined(CONFIG_SOC_DCDC_DA1469X)
+static void configure_dcdc(void)
+{
+#if defined(CONFIG_SOC_DCDC_DA1469X_1V8)
+	DCDC->DCDC_V18_REG |= DCDC_DCDC_V18_REG_DCDC_V18_ENABLE_HV_Msk;
+	DCDC->DCDC_V18_REG &= ~DCDC_DCDC_V18_REG_DCDC_V18_ENABLE_LV_Msk;
+#else
+	DCDC->DCDC_V18_REG &= ~DCDC_DCDC_V18_REG_DCDC_V18_ENABLE_HV_Msk;
+	DCDC->DCDC_V18_REG &= ~DCDC_DCDC_V18_REG_DCDC_V18_ENABLE_LV_Msk;
+#endif
+
+#if defined(CONFIG_SOC_DCDC_DA1469X_1V8P)
+	DCDC->DCDC_V18P_REG |= DCDC_DCDC_V18P_REG_DCDC_V18P_ENABLE_HV_Msk;
+	DCDC->DCDC_V18P_REG &= ~DCDC_DCDC_V18P_REG_DCDC_V18P_ENABLE_LV_Msk;
+#else
+	DCDC->DCDC_V18P_REG &= ~DCDC_DCDC_V18P_REG_DCDC_V18P_ENABLE_HV_Msk;
+	DCDC->DCDC_V18P_REG &= ~DCDC_DCDC_V18P_REG_DCDC_V18P_ENABLE_LV_Msk;
+#endif
+
+	DCDC->DCDC_VDD_REG |= DCDC_DCDC_VDD_REG_DCDC_VDD_ENABLE_HV_Msk;
+	DCDC->DCDC_VDD_REG |= DCDC_DCDC_VDD_REG_DCDC_VDD_ENABLE_LV_Msk;
+
+	DCDC->DCDC_V14_REG |= DCDC_DCDC_VDD_REG_DCDC_VDD_ENABLE_HV_Msk;
+	DCDC->DCDC_V14_REG |= DCDC_DCDC_VDD_REG_DCDC_VDD_ENABLE_LV_Msk;
+
+#if defined(CONFIG_PM)
+	/* Store DCDC registers to quickly restore DCDC state after wakeup */
+	dcdc_state.v18 = DCDC->DCDC_V18_REG;
+	dcdc_state.v18p = DCDC->DCDC_V18P_REG;
+	dcdc_state.vdd = DCDC->DCDC_VDD_REG;
+	dcdc_state.v14 = DCDC->DCDC_V14_REG;
+	dcdc_state.ctrl1 = DCDC->DCDC_CTRL1_REG;
+#endif
```
looks like a regulator, please take a look at regulator.h API

kasjer replied:
I looked at regulator API as requested.
Code you highlighted enables certain power rails according to Kconfig.
Regulator API covers a lot of functionality that is not available in this MCU case.
Code in soc.c uses similar approach as can be found in `nrf52/soc.c` were DCDC is configured as Kconfig dictates.
Please reconsider your change request.

gmarull replied:
Please explain why regulator API is not suitable. There are regulators for e.g. RPi, NXP VREF, etc. nRF code predates regulators, it may need to be revisited.

kasjer replied:
@gmarull, two commits for DCDC configuration removed from this PR. This can be later added as regulators.
Power management functionality implemented in this PR does not require configurability of DCDC.

gmarull commented on soc/arm/renesas_smartbond/da1469x/soc.h:
```diff
@@ -13,6 +13,16 @@ extern "C" {
+void z_smartbond_wakeup(void);
+void z_smartbond_restore_state(void);
+
+#define VDD_LEVEL_0V9		0
+#define VDD_LEVEL_1V2		3
+
+void z_smartbond_vdd_level_set(uint8_t vdd_level);
```
same

andrzej-kaczmarek replied:
can we just keep code that does *basic* settings to initialize power rails on init? we can full configuration via regulators in separate PR but we need time to get familiar with that API.

gmarull replied:
No if it fits the regulator API. Sometimes things require more effort. :-)

ioannis-karachalios commented on soc/arm/renesas_smartbond/da1469x/soc.c:
```diff
@@ -42,6 +53,50 @@ void sys_arch_reboot(int type)
+	POWER_CTRL_REG_SET(LDO_1V8P_ENABLE, 1);
+
+	/* Configure VDD to 1.2V if PLL is enabled, 0.9V otherwise. */
+	if (DT_NODE_HAS_STATUS(DT_NODELABEL(pll), okay)) {
```
The VDD level configuration should be removed here. This is already handled by the clock control driver. Another reason is that changing the VDD level should always be accompanied with updating the flash controller's pipe clock delay. Otherwise the controller will become unstable and this should end up raising ARM exceptions. The issue can be reproduced when disabling PLL (by default should be enabled in da1469x_dk_pro overlay).

kasjer replied:
@gmarull Added regulator driver that is now used
@ioannis-karachalios soc.c now does not set VDD voltage in init. It is done in regulator devices. Only when clock control requests PLL voltage is increased.

@ioannis-karachalios This code can be tested with regulator shell (preferable with #62190 applied). I noticed that when DCDC is selected for V18 rail and this rail is configured to 1.2V instead of default 1.8V. Rail is no powered during active mode and only is powered when device goes to sleep when LDO_RET is used instead. When rail is set to 1.8V LDO and DCDC work correct in active mode.

ioannis-karachalios replied:
@kasjer, as for setting VDD voltage, I knew that regulator devices would be introduced and so the existing implementation would no longer be valid. I just mentioned it, so you are aware of VDD update dependencies.  Next week, I will test the code with regulator shell and let you know if what you observe for V18 rail is expected or not.  Thanks for the update.

ioannis-karachalios replied:
@kasjer as for V18 rail not being able to generate 1.2V, is `DCDC_V18_REG[DCDC_V18_ENABLE_LV]` bit field set? Otherwise, low voltages cannot be generated (default value should be zero) which should also be reflected in the default device tree property values:

```
 v18: V18 {
     ...
    regulator-initial-mode =
	<DA1469X_MODE_DCDC_HIGH_BATTERY_VOLTAGE_ON>;
 };
```

kasjer replied:
@ioannis-karachalios maybe I had something set incorrect. With current code I don't see this anomaly any more. V18 when set to 1.2V works OK with DCDC and LDO in active and sleep mode.

ioannis-karachalios replied:
Thanks for the update. I just confirmed internally that `DCDC_V18_ENABLE_LV` bit field is not required for V18 rail to generate 1.2V

ioannis-karachalios commented on drivers/regulator/regulator_da1469x.c:
```diff
@@ -0,0 +1,461 @@
+};
+
+static const uint32_t voltages_v30[] = {
+	3000000, 3450000, 3300000, 3300000
```
V30 rail does not longer support 3.45 voltages. `3450000` option should be removed.

gmarull replied:
please ping me when regulator driver is ready

kasjer replied:
> V30 rail does not longer support 3.45 voltages. `3450000` option should be removed.

![da14695_regulators](https://github.com/zephyrproject-rtos/zephyr/assets/23063648/03454f14-e198-4694-ba70-eb2d5234daf3)
Devices that I have do support 3.45V as specified in datasheet.

kasjer replied:
@gmarull it's ready

ioannis-karachalios replied:
Maybe you have a TSMC version which is no longer available on the market. Here is an excerpt of POWER_CTRL_REG from the latest datasheet, that is version 3.3.
<img width="268" alt="V30_LEVEL" src="https://github.com/zephyrproject-rtos/zephyr/assets/13284615/a5444963-a988-4f85-9604-a8e3f2d23286">

ioannis-karachalios replied:
The same goes for the VDD sleep voltages, 0.95V and 1V are no longer supported in GF:
<img width="690" alt="GF_VDD_SLEEP_VOLTAGES" src="https://github.com/zephyrproject-rtos/zephyr/assets/13284615/414fb16c-c3ef-4457-9a1a-5d94a143abbf">
In contrast to TSMC variant:
<img width="509" alt="TSMC_VDD_SLEEP_VOLTAGES" src="https://github.com/zephyrproject-rtos/zephyr/assets/13284615/d89dec27-b4e5-44f4-b8f9-97303ae54786">
@kasjer The datasheet should reflect the GF silicon variant and not TSMC.

kasjer replied:
TSMC voltages removed as requested

ioannis-karachalios replied:
Thanks for the update

kasjer replied:
@gmarull I think you can take a look now

gmarull commented on drivers/regulator/Kconfig.da1469x:
```diff
@@ -0,0 +1,19 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config REGULATOR_DA1469X
+	bool "GPIO-controlled regulators"
```
this needs a better prompt

kasjer replied:
fixed

gmarull commented on drivers/regulator/Kconfig.da1469x:
```diff
@@ -0,0 +1,19 @@
+	default y
+	depends on DT_HAS_RENESAS_SMARTBOND_REGULATOR_ENABLED
+	help
+	  Enable the driver for GPIO-controlled regulators
```
same

kasjer replied:
fixed

gmarull commented on drivers/regulator/Kconfig.da1469x:
```diff
@@ -0,0 +1,19 @@
+if REGULATOR_DA1469X
+
+config REGULATOR_DA1469X_INIT_PRIORITY
+	int "Init priority"
```
same

kasjer replied:
fixed

gmarull commented on dts/arm/renesas/smartbond/da1469x.dtsi:
```diff
@@ -96,6 +97,57 @@
+			v30: V30 {
+				regulator-init-microvolt = <3000000>;
+				regulator-boot-on;
+				regulator-initial-mode = <
+					(DA1469X_MODE_SLEEP_LDO |
+					 DA1469X_MODE_LDO_VBAT_ON |
+					 DA1469X_MODE_LDO_VBUS_ON |
+					 DA1469X_MODE_V30_VBAT_CLAMP_ON)>;
```
what is this trying to configure? modes are not something to configure any parameters of a regulator, but "operating modes" (e.g. regulation schemes)

kasjer replied:
This sets that V30 rail can be powered from VBUS line (when available) or VBAT line, VBAT line without LDO for low battery condition, and that if should also switch to special clamp mode during sleep. For V30 there is no DCDC available so additional don't apply here.

gmarull replied:
sounds like these should be regular dt props, @aasinclair please review as well

aasinclair replied:
Agreed, should be handled as device specific properties, e.g:
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/regulator/raspberrypi%2Ccore-supply-regulator.yaml#L21-L24

kasjer replied:
@aasinclair @gmarull modes are now removed from driver, few properties added instead

gmarull commented on dts/arm/renesas/smartbond/da1469x.dtsi:
```diff
@@ -96,6 +97,57 @@
+					 DA1469X_MODE_LDO_VBAT_ON |
+					 DA1469X_MODE_LDO_VBUS_ON |
+					 DA1469X_MODE_V30_VBAT_CLAMP_ON)>;
+				regulator-v30-ref-bandgap;
```
please prefix vendor properties with `vnd,prop-name`

kasjer replied:
fixed

gmarull commented on include/zephyr/dt-bindings/regulator/da1469x.h:
```diff
@@ -0,0 +1,22 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_REGULATOR_DA1469X_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_REGULATOR_DA1469X_H_
+
+/*
+ * Set this mode if rail should be powered during sleep,
+ * mode applies to VDD, V18, V18P
+ */
+#define DA1469X_MODE_SLEEP_LDO					0x01
+/* Settings valid for VDD, V14, V18, V18P */
+#define DA1469X_MODE_DCDC_HIGH_BATTERY_VOLTAGE_ON		0x02
+#define DA1469X_MODE_DCDC_LOW_BATTERY_VOLTAGE_ON		0x04
+/* Settings valid for V30 */
+#define DA1469X_MODE_LDO_VBUS_ON				0x08
+#define DA1469X_MODE_LDO_VBAT_ON				0x10
+#define DA1469X_MODE_V30_VBAT_CLAMP_ON				0x20
```
see my previous comment, do not look as valid modes

kasjer replied:
DA1469X_MODE_DCDC_HIGH_BATTERY_VOLTAGE_ON means that rail should used DCDC when VBAT level is consider high, DA1469X_MODE_DCDC_LOW_BATTERY_VOLTAGE_ON means that DCDC can be used below this level. If both bits are 0 rail will not use DCDC.
DA1469X_MODE_SLEEP_LDO means that rail is powered during sleep.

gmarull reviewed (changes requested):
I'd suggest to submit a separate PR for the regulator driver

aasinclair commented on drivers/regulator/regulator_da1469x.c:
```diff
@@ -0,0 +1,451 @@
+		.rail = rail_id,                                               \
+		.flags = (DT_PROP_OR(inst, renesas_regulator_v30_clamp, 0) *   \
+			  DA1469X_MODE_V30_VBAT_CLAMP_ON) |                    \
+			 (DT_PROP_OR(inst, renesas_regulator_v30_vbus, 0) *    \
+			  DA1469X_MODE_LDO_VBUS_ON) |                          \
```
For a boolean you don't need OR, so I suggest:

```
DT_PROP(inst, renesas_regulator_v30_vbus) * DA1469X_MODE_LDO_VBUS_ON) 
```

kasjer replied:
done, thanks for a hint

aasinclair commented on drivers/regulator/regulator_da1469x.c:
```diff
@@ -0,0 +1,451 @@
+	return regulator_common_init(dev, 0);
+}
+
+#define REGULATOR_DA1469X_DEFINE(inst, id, rail_id)                            \
```
could you change the first argument from `inst` to `node`?

kasjer replied:
done

kasjer commented:
Force pushed after rebase to get rid of build conflicts

gmarull commented on drivers/regulator/regulator_da1469x.c:
```diff
@@ -0,0 +1,451 @@
+	862000, 828000, 798000, 706000
+};
+
+static const uint32_t voltages_vdd[] = {
+	900000, 1000000, 1100000, 1200000
+};
+
+static const uint32_t voltages_vdd_sleep[] = {
+	750000, 800000, 850000, 900000,
+};
+
+static const uint32_t voltages_v14[] = {
+	1200000, 1250000, 1300000, 1350000,
+	1400000, 1450000, 1500000, 1550000
+};
+
+static const uint32_t voltages_v30[] = {
+	3000000, 3300000
+};
+
+static const uint32_t voltages_v18[] = {
+	1200000, 1800000
+};
```
any reason to not use linear range APIs? Please, submit regulator driver as a new PR to facilitate reviewing

kasjer replied:
Yes, clamp is still not linear, V30 was non-liner till 3.45 was remove per @ioannis-karachalios request.
Voltages are selected by indices in registers since at lease one is non-linear all are treated the same with district values kept in code.
If Renesas decides yet again that V30 (or other rail) can have other values code will not require extensive change.

Separate PR for regulators #65226

gmarull commented on soc/arm/renesas_smartbond/da1469x/soc.c:
```diff
@@ -6,6 +6,8 @@
 #include <zephyr/init.h>
 #include <zephyr/linker/linker-defs.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/regulator.h>
```
where is regulator used?

kasjer replied:
unused header removed now

andrzej-kaczmarek commented:
@gmarull can we get final review on this one? All comments were already addressed.

gmarull commented on soc/arm/renesas_smartbond/da1469x/power.c:
```diff
@@ -0,0 +1,212 @@
+
+		/* PD_SYS will not be disabled here until we enter deep sleep - don't wait */
+		if (!da1469x_pd_release_nowait(MCU_PD_DOMAIN_SYS)) {
+			__DSB();
```
please, use barrier API

kasjer replied:
fixed

gmarull commented on soc/arm/renesas_smartbond/da1469x/soc.c:
```diff
@@ -31,13 +35,53 @@
 	NVIC_SystemReset();
 }
 
+void z_smartbond_restore_dcdc_state(void)
+{
+	/*
+	 * Enabling DCDC while VBAT is below 2.5 V renders system unstable even
+	 * if VBUS is available. Enable DCDC only if VBAT is above minimal value.
+	 */
+	if (CRG_TOP->ANA_STATUS_REG & CRG_TOP_ANA_STATUS_REG_COMP_VBAT_HIGH_Msk) {
+#if defined(CONFIG_PM)
+		DCDC->DCDC_V18_REG = dcdc_state.v18;
+		DCDC->DCDC_V18P_REG = dcdc_state.v18p;
+		DCDC->DCDC_VDD_REG = dcdc_state.vdd;
+		DCDC->DCDC_V14_REG = dcdc_state.v14;
+		DCDC->DCDC_CTRL1_REG = dcdc_state.ctrl1;
+#endif
+	}
+}
+
+void z_smartbond_store_dcdc_state(void)
+{
+#if defined(CONFIG_PM)
+	dcdc_state.v18 = DCDC->DCDC_V18_REG;
+	dcdc_state.v18p = DCDC->DCDC_V18P_REG;
+	dcdc_state.vdd = DCDC->DCDC_VDD_REG;
+	dcdc_state.v14 = DCDC->DCDC_V14_REG;
+	dcdc_state.ctrl1 = DCDC->DCDC_CTRL1_REG;
+#endif
+}
```
why is this touching DC/DC without regulator API?

kasjer replied:
When device goes to sleep some register values (including DC/DC) are lost. This code stores settings on sleep and restores them on wakeup. Foe each rail that do have DCDC option only DCDC register values are lost. Other configuration (i.e. voltages) are kept intact. Sleep in Smartbond also looses CPU registers that are restored in PM code. To make things fast there is no call to regulator API since no change in regulator is expected.

gmarull replied:
this should be handled by device drivers, not in a random file that does not even respect DT (what is "DCDC"?)

andrzej-kaczmarek replied:
This has nothing to do with DT or device driver. Extended sleep on DA1469x effectively powers off ARM core and we need to store some registers to restore them quickly on wakeup. If we pass this through all the subsystems in Zephyr instead of just restoring from local variables, the wake up process will be unnecessarily long for no reason.

andrzej-kaczmarek replied:
because we only need to do this in case ARM core and was shut down, otherwise those registers are not affected - PM events do not know if that was the case

kasjer commented on drivers/timer/smartbond_timer.c:
```diff
@@ -0,0 +1,184 @@
+
+#define COUNTER_SPAN BIT(24)
+#define CYC_PER_TICK (sys_clock_hw_cycles_per_sec() / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+#define TICK_TO_CYC(tick) ((tick) * sys_clock_hw_cycles_per_sec() / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
```
code was based on what nrf_timer_timer.c has.
Now it is changed to use conversion from  time_units.h

andrzej-kaczmarek commented:
@gmarull could you please follow up on latest comments/fixes?

andrzej-kaczmarek commented:
@gmarull we analyzed possibility of restoring DCDC registers in regulators driver and it simply doesn't make any sense in this case.

This is *not* the case where device has to be prepared for sleep and then resumed. Those registers are lost only if ARM core is turned off which doesn't happen on every suspend and only SoC code can tell the difference.

If we move this to regulators driver then it will be either of 2 cases:
1. We use global hook for PM events. Handling those registers is not per-device and thus still doesn't respect DT as mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/61857#discussion_r1423724622
2. We use per-device hook for PM events. Each rail restores it's DCDC state but also has to care about global DCDC state because it has to be enabled first (i.e. each rail also always needs to check if DCDC was already enabled by other rail). Also this implies that PM_DEVICE has to be selected by PM as otherwise restoring DCDC state won't work properly - I don't see any other SoC that does it.

In any case, those hooks will "restore" DCDC registers state (in 2nd case registers will be accessed multiple times) even if not needed so not really smart in terms of power management.

^^ @aasinclair can you comment on regulators pov?

kasjer committed:
drivers: regulator: Add power management to DCDC

This change restore DCDC configuration after system resumes.

When CONFIG_PM_DEVICE is enabled each of the four rails that
support DCDC handle resume in regulator_da1469x_pm_action function.

Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

andrzej-kaczmarek commented:
If either 1 or 2 can be approved here anyway, we'll go with that for now. If not, please give us some idea what should be done here since we're running out of options.

gmarull commented:
> If either 1 or 2 can be approved here anyway, we'll go with that for now. If not, please give us some idea what should be done here since we're running out of options.

Please take a look to https://github.com/zephyrproject-rtos/zephyr/issues/59194 and https://github.com/zephyrproject-rtos/zephyr/pull/60998. The idea is simple, if your device looses context due to a system PM transition, it's device responsibility to bring its state back, not code in a random soc/ file. That PR proposes to modify the existing notification mechanism to allow subscriptions to specific states. Unfortunately, PM in Zephyr is a mess, but that doesn't justify taking shortcuts.

andrzej-kaczmarek commented:
> > If either 1 or 2 can be approved here anyway, we'll go with that for now. If not, please give us some idea what should be done here since we're running out of options.
> 
> Please take a look to #59194 and #60998. The idea is simple, if your device looses context due to a system PM transition, it's device responsibility to bring its state back, not code in a random soc/ file. That PR proposes to modify the existing notification mechanism to allow subscriptions to specific states. Unfortunately, PM in Zephyr is a mess, but that doesn't justify taking shortcuts.

Those PRs don't make any difference here because there's no separate power state where DCDC registers are lost vs. power state where they are preserved. This only depends on SoC when it enters extended sleep mode. No matter if we subscribe to single PM state transition or all of them, the driver can still only restore registers on every transition which in many cases is just redundant. I don't really know how I can make it any more clear.

We're not trying to make any shortcuts here, we just want to solve use-case specific to this particular SoC in a few simple lines of code which is not possible with current PM subsystem. I don't understand why this PR needs to wait until PM subsystem can handle that. There are other SoCs that take huge shortcuts with PM implementation and somehow it's not a big deal, see e.g. https://github.com/zephyrproject-rtos/zephyr/pull/45357.

gmarull commented:
> > > If either 1 or 2 can be approved here anyway, we'll go with that for now. If not, please give us some idea what should be done here since we're running out of options.
> > 
> > 
> > Please take a look to #59194 and #60998. The idea is simple, if your device looses context due to a system PM transition, it's device responsibility to bring its state back, not code in a random soc/ file. That PR proposes to modify the existing notification mechanism to allow subscriptions to specific states. Unfortunately, PM in Zephyr is a mess, but that doesn't justify taking shortcuts.
> 
> Those PRs don't make any difference here because there's no separate power state where DCDC registers are lost vs. power state where they are preserved. This only depends on SoC when it enters extended sleep mode. No matter if we subscribe to single PM state transition or all of them, the driver can still only restore registers on every transition which in many cases is just redundant. I don't really know how I can make it any more clear.
> 

What you're doing in pm_state_set after WFI, should actually be in post ops (these are really post ops, not even related to entering a PM state). These post-ops, in particular restoring DC/DC state (which has a driver I believe), should be handled by the driver itself using something like what is proposed in #60998, code that is run at the same point but in a distributed manner (ie drivers get called when exiting a power state). This way, there's no need to add device related code in soc/ for things that have a driver.

> We're not trying to make any shortcuts here, we just want to solve use-case specific to this particular SoC in a few simple lines of code which is not possible with current PM subsystem. I don't understand why this PR needs to wait until PM subsystem can handle that. There are other SoCs that take huge shortcuts with PM implementation and somehow it's not a big deal, see e.g. #45357.

My -1 stands until this is proposed with a better solution. Sorry but no to more vendor random stuff in soc/.

andrzej-kaczmarek commented:
We will fix this code once Zephyr has proper interfaces. That mentioned PR is stalled so I don't see it happening any time soon and the fact that we cannot merge essential functionality to our SoC for almost half a year because of missing functionality in Zephyr, while other SoCs already have code that almost workarounds entire PM subsystem, is just inconceivable.

gmarull commented:
> We will fix this code once Zephyr has proper interfaces. That mentioned PR is stalled so I don't see it happening any time soon and the fact that we cannot merge essential functionality to our SoC for almost half a year because of missing functionality in Zephyr, while other SoCs already have code that almost workarounds entire PM subsystem, is just inconceivable.

Then please, work on improving Zephyr capabilities. It’s easy to drop vendor stuff in soc/, or to just implement drivers. Infrastructure, new APIs, etc. takes more time, but that’s the way it is.

kasjer commented:
@gmarull DCDC access moved to regulators.
code can be compiled with PM_DEVICE or without it with just PM

gmarull commented on drivers/regulator/regulator_da1469x.c:
```diff
@@ -413,3 +468,36 @@ static int regulator_da1469x_init(const struct device *dev)
+
+int regulators_da1469x_pm_init(void)
+{
+	pm_notifier_register(&regulators_da1469x_pm_notifier);
```
this can be done by the driver itself, init function. Since the driver seems to be for a singleton device, you can avoid the usage of the generic foreach instantiation (there are a few driver in tree doing this). This way, the fact that the notifier API can't pass context (would be needed for multiinstance devices) doesn't matter.

kasjer replied:
suggestion applied

gmarull commented on drivers/regulator/regulator_da1469x.c:
```diff
@@ -369,6 +390,37 @@ static int regulator_da1469x_init(const struct device *dev)
 	return regulator_common_init(dev, 0);
 }
 
+#if IS_ENABLED(CONFIG_PM_DEVICE)
+static int regulator_da1469x_pm_action(const struct device *dev,
+				       enum pm_device_action action)
+{
+	const struct regulator_da1469x_config *config = dev->config;
+	int ret = 0;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		if (config->desc->dcdc_register) {
+			*config->desc->dcdc_register = *config->desc->dcdc_register_shadow;
+			if ((CRG_TOP->ANA_STATUS_REG & CRG_TOP_ANA_STATUS_REG_COMP_VBAT_HIGH_Msk) &&
+			    (*config->desc->dcdc_register_shadow & DCDC_REQUESTED)) {
+				DCDC->DCDC_CTRL1_REG = dcdc_state.ctrl1;
+			}
+		}
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		/*
+		 * Shadow register is saved on each regulator API call, there is no need
+		 * to save it here.
+		 */
+		break;
+	default:
+		ret = -ENOTSUP;
+	}
+
+	return ret;
+}
+#endif
```
do you really need this if notifier bit is in place?

kasjer replied:
removed

gmarull replied:
where is this removed?

gmarull commented on soc/arm/renesas_smartbond/da1469x/power.c:
```diff
@@ -0,0 +1,141 @@
+			slept = z_smartbond_sleep();
+		}
+
+		if (slept) {
+			if (!IS_ENABLED(CONFIG_WDT_SMARTBOND)) {
+				/* Watchdog is always resumed when PD_SYS is turned off, need to
+				 * freeze it again if there's no one to feed it.
+				 */
+				GPREG->SET_FREEZE_REG = GPREG_SET_FREEZE_REG_FRZ_SYS_WDOG_Msk;
+				SYS_WDOG->WATCHDOG_REG = SYS_WDOG_WATCHDOG_REG_WDOG_VAL_Msk;
+			}
+
+			da1469x_pd_acquire(MCU_PD_DOMAIN_SYS);
+
+			if (z_smartbond_is_wakeup_by_jtag()) {
+				wait_for_jtag = 1;
+			}
+
+			if (DT_SAME_NODE(DT_PROP(DT_NODELABEL(sys_clk), clock_src),
+					 DT_NODELABEL(xtal32m))) {
+				da1469x_clock_sys_xtal32m_switch_safe();
+			}
+		} else {
+			da1469x_pd_acquire_noconf(MCU_PD_DOMAIN_SYS);
+		}
+
+		break;
```
these seem to be post ops

andrzej-kaczmarek replied:
How can we pass `slept` flag to post-ops?

gmarull replied:
static variable?

kasjer replied:
moved to post_ops

gmarull commented on soc/arm/renesas_smartbond/da1469x/soc.c:
```diff
@@ -15,6 +16,9 @@
 #include <da1469x_trimv.h>
 #include <cmsis_core.h>
 
+#define LOG_LEVEL CONFIG_SOC_LOG_LEVEL
+LOG_MODULE_REGISTER(soc);
```
`LOG_MODULE_REGISTER(soc, CONFIG_SOC_LOG_LEVEL);`

gmarull commented on soc/arm/renesas_smartbond/da1469x/soc.h:
```diff
@@ -13,6 +13,11 @@ extern "C" {
 
 #include <DA1469xAB.h>
 
+int z_smartbond_sleep(void);
+void z_smartbond_wakeup_handler(void);
+void z_smartbond_wakeup(void);
+void z_smartbond_restore_state(void);
```
please do not pollute soc.h which is included almost everywhere (due to CMSIS design), create a dedicated header.

kasjer replied:
moved to smartbond_sleep.h

andrzej-kaczmarek commented:
@gmarull ping

blauret commented:
@gmarull ping

ioannis-karachalios commented on soc/arm/renesas_smartbond/da1469x/power.c:
```diff
@@ -0,0 +1,145 @@
+	ARG_UNUSED(state);
+	ARG_UNUSED(substate_id);
+
+	if (slept) {
```
This part should be executed only when `z_smartbond_is_sleep_allowed` returns true and so `da1469x_pd_release_nowait` gets executed in `pm_state_set` PM handler. Otherwise, every call to `da1469x_pd_acquire_noconf` or `da1469x_pd_acquire` should result in keep increasing the corresponding internal counter and thus preventing PD_SYS from being turned off.

kasjer replied:
@ioannis-karachalios now slept is only set in one place and reset when after true wake up.
Thanks for review.

kasjer replied:
@ioannis-karachalios you are right let me fix this...

ioannis-karachalios commented on soc/arm/renesas_smartbond/da1469x/power.c:
```diff
@@ -0,0 +1,145 @@
+			da1469x_clock_sys_xtal32m_switch_safe();
+		}
+	} else {
+		da1469x_pd_acquire_noconf(MCU_PD_DOMAIN_SYS);
```
I think this should still not work as `da1469x_pd_acquire_noconf` will get called even if `z_smartbond_is_sleep_allowed` returns false and breaks the while loop above (thus, keep increasing the PD_SYS reference counter). Isn't it?

kasjer replied:
@ioannis-karachalios I change code according to your suggestion, except I though that maybe `z_smartbond_is_sleep_allowed()` can return different value when called again (interrupt raised in the meantime) so instead of calling this function I keep track if PD_SYS was released.

ioannis-karachalios replied:
I agree that tracking `PD_SYS` is a more concrete approach. It should work as expected now.

andrzej-kaczmarek commented:
@gmarull ping

ioannis-karachalios commented on soc/arm/renesas_smartbond/da1469x/power.c:
```diff
@@ -0,0 +1,153 @@
+			wait_for_jtag = 1;
+		}
+
+		if (DT_SAME_NODE(DT_PROP(DT_NODELABEL(sys_clk), clock_src),
```
@kasjer I want to clarify this - system PM will first call `pm_resume_devices` and then `pm_state_exit_post_ops` where the system clock is switched from `RC32M` --> `XTAL32M` (after exiting the sleep state). It might happen that a device driver requests PLL from within its registered PM callback. In such a case, an assertion should be thrown since PLL is requested while XTAL32M has yet to be enabled (which is not allowed). To my view there are three approaches to overcome this:
 - Enabling XTAL32M is moved to `pm_state_set` right after `sleep_state = SLEEP_STATE_SLEPT` (before calling `pm_resume_devices`)
 - A device driver should first check that XTAL32M is up and running before switching to PLL (this should be documented) 
 - Following to enabling XTAL32M, we also track whether the device was running @ PLL so the last selected clock is re-applied. In doing so, a device driver should not have to check the current clock that might have been changed due to the device entering the sleep state.

To my view, the last option is the best approach.

kasjer replied:
@ioannis-karachalios we could fix PLL handling later. I only tested without PLL, since.

ioannis-karachalios replied:
OK. We should revisit this part at a next round.

ioannis-karachalios commented on soc/arm/renesas_smartbond/da1469x/power.c:
```diff
@@ -74,6 +139,8 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 			break;
 		}
 
+		gpio_latch();
```
GPIO latching should not take place here. Peripheral blocks powered by `PD_SYS`, that should be LCDC, DMA, CRYPTO etc, should first acquire PD_SYS (by calling `da1469x_pd_acquire_noconf`) before performing an operation/transaction. That should result in increasing the associated reference counter and so `da1469x_pd_release_nowait` (called below) should return zero. In doing so, the system should enter the WFI state instead of the sleep state. Typically, an interrupt designating the end of a transaction/operation should exit the device from the idle state. With the current implementation driver's signals are freezed. GPIO latching/un-latching shoud be handled as follow:

```
		if (!da1469x_pd_release_nowait(MCU_PD_DOMAIN_SYS)) {
			barrier_dsync_fence_full();
			__WFI();
		} else {
			gpio_latch();
			da1469x_pdc_ack_all_m33();
			if (z_smartbond_sleep()) {
				sleep_state = SLEEP_STATE_SLEPT;

				if (DT_SAME_NODE(DT_PROP(DT_NODELABEL(sys_clk), clock_src),
					DT_NODELABEL(xtal32m))) {
					da1469x_clock_sys_xtal32m_switch_safe();
				}
			}
			gpio_unlatch();
		} 
```

kasjer replied:
@ioannis-karachalios change applied

gmarull commented on dts/arm/renesas/smartbond/da1469x.dtsi:
```diff
@@ -16,6 +16,15 @@
 		zephyr,flash-controller = &flash_controller;
 	};
 
+	power-states {
+		standby: standby {
+			compatible = "zephyr,power-state";
+			power-state-name = "standby";
+			min-residency-us = <500>;
+			exit-latency-us = <100>;
+		};
+	};
```
power-states needs to be child of cpus node by convention

kasjer replied:
moved according to convention

gmarull commented on soc/renesas/smartbond/da1469x/CMakeLists.txt:
```diff
@@ -8,5 +8,7 @@ zephyr_linker_sources(
 zephyr_library()
 zephyr_library_sources(soc.c)
 zephyr_include_directories(.)
+zephyr_library_sources_ifdef(CONFIG_PM power.c)
+zephyr_library_sources_ifdef(CONFIG_PM sleep.S)
```
zephyr_library_sources_ifdef(CONFIG_PM power.c sleep.S)

kasjer replied:
sleep.S removed

gmarull commented on soc/renesas/smartbond/da1469x/power.c:
```diff
@@ -0,0 +1,152 @@
+#include <zephyr/pm/pm.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/barrier.h>
+#include <soc.h>
```
soc.h is not needed here, please, include the necessary headers (maybe cmsis core for nvic?)

gmarull commented on soc/renesas/smartbond/da1469x/power.c:
```diff
@@ -0,0 +1,152 @@
+	SLEEP_STATE_SLEPT,
+} sleep_state;
+
+__weak void pm_state_set(enum pm_state state, uint8_t substate_id)
```
no __weaks, this is no longer supported

kasjer replied:
__weaks removed

gmarull commented on soc/renesas/smartbond/da1469x/power.c:
```diff
@@ -0,0 +1,152 @@
+	}
+}
+
+__weak void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
```
same

gmarull commented on soc/renesas/smartbond/da1469x/power.c:
```diff
@@ -0,0 +1,152 @@
+	ARG_UNUSED(substate_id);
+
+	if (sleep_state == SLEEP_STATE_SLEPT) {
+		if (!IS_ENABLED(CONFIG_WDT_SMARTBOND)) {
+			/* Watchdog is always resumed when PD_SYS is turned off, need to
+			 * freeze it again if there's no one to feed it.
+			 */
+			GPREG->SET_FREEZE_REG = GPREG_SET_FREEZE_REG_FRZ_SYS_WDOG_Msk;
+			SYS_WDOG->WATCHDOG_REG = SYS_WDOG_WATCHDOG_REG_WDOG_VAL_Msk;
+		}
```
this seems like an action the watchdog driver should perform using eg. a notification mechanism when the system wakes up. I already expressed my concerns on spreading code across soc/ and drivers/. You're also bypassing DT to obtain device addresses (SYS_WDOG, GPREG).

kasjer replied:
code removed

gmarull commented on soc/renesas/smartbond/da1469x/sleep.S:
```diff
@@ -0,0 +1,265 @@
+
+    .equ CLK_AMBA_REG,              0x50000000
+    .equ RESET_STAT_REG,            0x500000BC
+    .equ NVIC_BASE,                 0xE000E100
```
isn't that address available from DT?

kasjer replied:
sleep.S removed

gmarull commented on soc/renesas/smartbond/da1469x/smartbond_sleep.h:
```diff
@@ -0,0 +1,23 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _SLEEP__H_
+#define _SLEEP__H_
```
align guarding with header names

kasjer replied:
header removed

gmarull commented on soc/renesas/smartbond/da1469x/power.c:
```diff
@@ -22,6 +22,18 @@ static int pdc_idx_combo;
 static int pdc_idx_sw_trigger;
 static bool wait_for_jtag;
 
+#define GPIO0_NGPIOS DT_PROP(DT_NODELABEL(gpio0), ngpios)
+#define GPIO1_NGPIOS DT_PROP(DT_NODELABEL(gpio1), ngpios)
+
+struct gpio_state_data {
+	uint32_t data0;
+	uint32_t mode0[GPIO0_NGPIOS];
+	uint32_t data1;
+	uint32_t mode1[GPIO1_NGPIOS];
+};
+
```
again, stuff that should be in driver/ code in soc/. Please, find a better solution. If Zephyr notification mechanism isn't sufficient, you can improve it or look at suggested improvements that are on PR.

kasjer replied:
latching gpio moved to gpio driver

gmarull reviewed (changes requested):
Still has some important issues.

kasjer committed:
manifest: hal_renesas: update module

Update hal_renesas to include sleep code

Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

gmarull reviewed (changes requested):
please, do not put Zephyr code in HALs.

kasjer commented:
while there was no Zephyr code in hal some includes were pulled from zephyr tree for asm macros, now it's removed

andrzej-kaczmarek commented:
@gmarull ping

gmarull commented on drivers/gpio/gpio_smartbond.c:
```diff
@@ -318,12 +418,16 @@ static const struct gpio_driver_api gpio_smartbond_drv_api_funcs = {
 		.wkup_regs = (volatile struct gpio_smartbond_wkup_regs *)		\
 						DT_INST_REG_ADDR_BY_NAME(id, wkup),	\
 		.wkup_trig_select = id,							\
+		GPIO_PM_DEVICE_CFG(ngpios, DT_PROP(DT_NODELABEL(gpio##id), ngpios))	\
```
this is wrong, there's no guarantee instance '0' will be 'gpio0'.

kasjer replied:
fixed, `gpio_smartbond_saved_state_<n>` generated for each gpio port if needed when **PM_DEVICE** is enabled

gmarull commented on drivers/gpio/gpio_smartbond.c:
```diff
@@ -318,12 +413,16 @@ static const struct gpio_driver_api gpio_smartbond_drv_api_funcs = {
 		.wkup_regs = (volatile struct gpio_smartbond_wkup_regs *)		\
 						DT_INST_REG_ADDR_BY_NAME(id, wkup),	\
 		.wkup_trig_select = id,							\
+		GPIO_PM_DEVICE_CFG(.ngpios, DT_PROP(DT_NODELABEL(gpio##id), ngpios))	\
```
this is still wrong. You can't use nodelabels in instance-based macros. Instance `0` may be `gpio1`.

kasjer commented:
Last force-push is just rebase to main

swift-tk commented:
I believe this PR broke "select GPIO"
See https://github.com/zephyrproject-rtos/zephyr/issues/72556
@andrzej-kaczmarek @carlescufi
